### DISCUSSION TITLE
How to use postgres JSONB subscripting to assign values in update

### DISCUSSION THREAD

**Discussion by rasca** [sqlalchemy/sqlalchemy#12033] [Usage Questions]
I need to update several thousand of sub keys in a jsonb field in postgreqsl.

I've tried with nested `jsonb_set` but I quickly reach Python's nesting limits.

In postgresql doing this works great:
    
    UPDATE table
    SET
        json_field['a']='"value"',
        json_field['b']='"value"'
        ...

for several thousand keys.

But I don't know how to do this in sqlalchemy.

A related issue is: https://github.com/sqlalchemy/sqlalchemy/issues/10927# which tracks moving from `json_field -> 'a'` to `json_field['a']`.

This is my current code but I get this error:

```
        sql_updates = {literal_column(f"LTM.state[{']['.join(path)}]"): cast(value, JSONB) for path, value in updates}
        with get_session() as session:
            session.execute(
                update(LTM)
                .where(LTM.id == ltm_id)
                .values(
                    **sql_updates
                )
            )
            session.commit()
```
Error:
```
TypeError: keywords must be strings
```

Apart from trying with `column_literal` I've tried with `text` but it expects strings as keys.

**Comment by CaselIT** [comment]
Hi,

That's a python error, not an sqlalchemy one. I suggest checking there it's causing the error and avoid doing what's causing it

**Comment by rasca** [comment]
That error appears when I use anything but strings in the keys of

```
                .values(
                    **sql_updates
                )
 ```

I'm asking how to use raw sql in the left side of an update's assingment.

Thanks!

**Reply by CaselIT** [reply]
have you tried not using `**`?

**Reply by rasca** [reply]
I need to dynamically create lots of nested sub keys updates. I don't know how to do this:

    UPDATE table
    SET
        json_field['a']['c']='"value"',
        json_field['b']['d']='"value"'
        ...

For several thousand keys.

Using merge `||` doesn't work because it's nested subkeys and not just first level ones.

Using `jsonb_set` doesn't work because it doesn't support the (sqlalchemy side) so many nested objects.

In plan SQL it works with the above code but I cannot get sqlalchemy to write it. Is there something else other than `.values()` I can use to construct this?

**Reply by CaselIT** [reply]
What's the new issue you are getting now that you fixed the type error?

**Reply by rasca** [reply]
If I use plain columns like `state` it works, but when I try to use `state['a']['b']` I can no longer do it with `values` e.g.:

`state=value1, state=value3` is valid Python
`state['a']['b']=value1, state['b']['c']=value3` is invalid Python and what I'm asking help to code

**Reply by CaselIT** [reply]
I know, I've already suggested how to fix: https://github.com/sqlalchemy/sqlalchemy/discussions/12033#discussioncomment-11062653

To be clearer: try using `.values(sql_updates)`

**Reply by zzzeek** [reply]
> If I use plain columns like `state` it works, but when I try to use `state['a']['b']` I can no longer do it with `values` e.g.:
> 
> `state=value1, state=value3` is valid Python `state['a']['b']=value1, state['b']['c']=value3` is invalid Python and what I'm asking help to code

you can control the rendering completely using a custom compiles directive for now

```py
import re

from sqlalchemy import column
from sqlalchemy import table
from sqlalchemy.dialects import postgresql
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql import operators
from sqlalchemy.sql.elements import BinaryExpression


@compiles(BinaryExpression, "postgresql")
def compile_jsonb_index(element, compiler, **kw):
    if element.operator is operators.json_getitem_op and isinstance(
        element.left.type, JSONB
    ):
        supertext = compiler.visit_binary(element, **kw)
        elems = re.match(r"\(?(\S+) -> (\S+)\)?", supertext)
        if elems:
            return f"{elems.group(1)}[{elems.group(2)}]"
        return supertext
    else:
        return compiler.visit_binary(element, **kw)


t = table("table", column("json_field", JSONB))
stmt = t.update().values(
    {t.c.json_field["a"]["b"]: "value", t.c.json_field["b"]["c"]: "value"}
)

print(stmt.compile(dialect=postgresql.dialect()))

```

**Comment by rasca** [comment]
I'm so sorry, in my raw SQL version I have it like that. Thanks a lot for the patience 🙏 


When using `column_literals` I get `sqlalchemy.exc.CompileError: Unconsumed column names: state['_status'],`

**Reply by CaselIT** [reply]
I'm not sure if there is a way of skipping that check.

@zzzeek do you have a suggestion here?

**Reply by zzzeek** [reply]
can I get some code to work with ?  what's `_status` for example?   what is the desired SQL ?

**Reply by zzzeek** [reply]
> I need to update several thousand of sub keys in a jsonb field in postgreqsl.

> I've tried with nested jsonb_set but I quickly reach Python's nesting limits.

What's being nested?   

```
UPDATE table
SET
    json_field['a']='"value"',
    json_field['b']='"value"'
    ...
```

would be

```
UPDATE table
SET json_field=jsonb_set(json_field, {"a": "value", "b": "value", ...})
```

```py
>>> from sqlalchemy.dialects.postgresql import JSONB
>>> from sqlalchemy import func, table, column
>>> t = table('table', column('json_field', JSONB))
>>> stmt = t.update().values(json_field=func.jsonb_set(t.c.json_field, {"a": "value", "b": "value", "c": "value"}))
>>> print(stmt)
UPDATE "table" SET json_field=jsonb_set("table".json_field, :jsonb_set_1)

```

the bound value here can be as big as you'd like, there's no "nesting" on the Python side
